### ISSUE TITLE
Handle long type parameter lists

### ISSUE THREAD

**Issue by emchristiansen** [scala-ide/scalariform#77]
First, thanks for developing Scalariform. I use it hundreds of times a day.

The issue: I code in an extremely functional style, and often have very long type parameter lists. For example, this is a function signature from my code:

def relativeBenchmark[E <% RuntimeConfig => ExperimentRunner[R], R <% RuntimeConfig => ExperimentSummary](experiment: E): Double

Scalariform, used from Eclipse, tries to put all the type parameters on one line, meaning the above signature would extend beyond the 80 character mark.

Ideally: Please detect when a type parameter list is long and distribute it across multiple lines. I like this format:

def relativeBenchmark[
  E <% RuntimeConfig => ExperimentRunner[R], 
  R <% RuntimeConfig => ExperimentSummary](experiment: E): Double

I'd be happy with: Don't format type parameter lists that extend across multiple lines.

**Comment by ochafik** [comment]
+1

The problem only gets worse with widespread `private[this]` usage. I'm not using scalariform in Scalaxy/Streams because I'm fond of the following style:

```
  private[this]
  def bar[A <: Foo : Bar,
          B : Baz : Bam]
         (x: A, y: B)
         (implicit z: Int)
         : Foo = ???
```
instead of:
```
  private[this] def bar[A <: Foo : Bar, B : Baz : Bam](x: A, y: B)(implicit z: Int): Foo = ???
```

**Comment by cvogt** [comment]
+1
Here is my favorite variant:
```

**Comment by sergius** [comment]
+1 
Long type parameter list suggested by @emchristiansen.
Any progress in this regard since 2013? :smile:

**Comment by hawkw** [comment]
I'd also really like to see this feature, and I'd be willing to open a PR to implement it myself. However, I could use a few pointers from a maintainer with regard to where to start in the codebase, if anyone's willing to point me in the right direction.

**Comment by godenji** [comment]
@hawkw 

At a glance it looks like a quick and dirty proof of concept could be achieved by disabling formatting [on this line](https://github.com/scala-ide/scalariform/blob/master/scalariform/src/main/scala/scalariform/formatter/TypeFormatter.scala#L55). i.e. change to `NoFormatResult`

Then [create an executable jar](https://github.com/scala-ide/scalariform#packaging-an-executable-jar) to test whether or not sample scala file containing various long type paramater lists leaves applicable code blocks unformatted.

If that works we can look into creating a new Preference Descriptor that allows users to opt-in to type parameter list formatting.

**Comment by godenji** [comment]
This is fixed in 0.2.4

To enable the below formatting style set either `AllowParamGroupsOnNewlines` or `allowParamGroupsOnNewlines` to true in your `build.sbt/.scalariform.conf` file. 
```
def foo[T]
  (a: A)
  (b: B)
  (implicit t: T)
```
Note that this does *not* automatically format sources to the desired format, but rather prevents the default behavior of collapsing param groups into a single line.

Also, type params can be placed on a newline, but unlike paren groups, a bracket group spanning multiple lines collapses to a single line (sorry, bracket handling code is seriously *hairy*).
